### PR TITLE
Phoebes branch

### PR DESCRIPTION
**Pull request by phoebeli23**
## In this PR

React app to display data from an Airtable goal tracking database

Link to Airtable: 
https://airtable.com/invite/l?inviteId=invMHujtD9ROC5cYZ&inviteToken=68d0143c05ec3ffe6eec4c82a5fdae7903c0d935828a3eff09cf1e0ca3c5599e

### 🖋  What I learned

- use Airtable's online interface
- connect React to Airtable via its API
- build with styled-components

### 🤳 Screenshots (if applicable)

Screen Shot 2020-10-03 at 11 22 59 PM

### Pop Quiz 

1. Why do tables need primary keys? What are the two conditions that need to be met for a column to be a primary key?
Primary keys are used identify records when defining schema. They must be unique and stable. 

2. Define relation. How were relations used in the goal tracker app? 
Relations describe how different tables are related to each other. For example, in the goal tracker app, a goal can have multiple updates, but an update can only have one goal. This is accomplished by the update field in the goal table, which links to multiple records in the update table. Similarly, the goal field in the update table links to a single goal record. 

3. Define foreign key. How were foreign keys used in the goal tracker app? 
A foreign key is a reference to a primary key from another table, used to establish relations. The link to [other table] fields are examples of foreign keys in this project. 

### 🐞 Questions

1. What exactly are "pages" in a table? We use .eachPage here, but I think our tables only have one page since we setGoals and setUpdates to the records parameter each time. Is my guess right?

Screen Shot 2020-10-03 at 11 48 20 PM

2. Do we usually use styled components or css modules?

3. I initially tried implementing this project in TS, but found it quite difficult and didn't think I would have enough time, since the API docs are written for JS. How do you go about figuring out the types of objects returned for APIs without docs specifically for TS?

### CODE REVIEW

File: src/components/Goal.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
+import StyledCheckBox from '../styles/StyledCheckbox';
+import StyledGoal from '../styles/StyledGoal';
+import StyledGoalDetails from '../styles/StyledGoalDetails';
```
Comment (aliceyezhao) [inline src/components/Goal.js:4]: Could condense into one import line

import {x, y, z} from 'path';

**Review by aliceyezhao — APPROVED** [review]
Looks awesome! Great work Phoebe :D

**Comment by micahtyong** [comment]
Awesome job with this Tiffany!

**Comment by micahtyong** [comment]
> What exactly are "pages" in a table? We use .eachPage here, but I think our tables only have one page since we setGoals and setUpdates to the records parameter each time. Is my guess right?

**From Ethan**: API results tend to be paged less for the client and more so that the server can deliver quick, lightweight API responses. If we send the response in chunks, we get some data to work with (the initial chunk) faster than it would take to perform a full scan and send all of that over the wire.

It’s often the case that you don’t need “all of the API results” anyway, so it’s smart to page from a UX perspective as well. In the few cases where you definitely don’t need to page, Airtable provides a .all convenience function that pages through everything automatically (effectively a full scan). See an example [here on 12:28](https://www.notion.so/calblueprint/Introduction-to-Airtable-js-96794d2f48c5476ebf08a6bf2ba9da50#99f555fc3e7e4db09d4dccfd3a737639). 

**TL;DR**: The page size isn’t based on local memory, it’s based on the maxRecords parameter, where the max for that parameter is 100, IIRC. They cap that to deliver fast, lightweight API responses that don’t break the bank from a compute perspective. 

> Do we usually use styled components or css modules?

It largely depends on the team! Ask Alice or Nick about this. Styled components have been a popular choice in the past, but many teams are choosing to import their own UI libraries this year. 

> I initially tried implementing this project in TS, but found it quite difficult and didn't think I would have enough time, since the API docs are written for JS. How do you go about figuring out the types of objects returned for APIs without docs specifically for TS?

Ah yes. I should have given a more explicit warning earlier. TypeScript and AirTable, for the version used in your project, are not very compatible. Last year, we had to define all the AirTable types ourselves in the [1951 coffee](https://github.com/calblueprint/1951coffee/tree/master/utils/airtable) codebase. However, the newest update of AirTable does have TypeScript compatibility!
